### PR TITLE
Updated C# and Html projection providers to allow doc removals.

### PR DESCRIPTION
**Pull request by NTaylorMullen**
- When a document is removed the Html/C# projected document providers now notify VSCode that the projected documents are empty. This works around the projected document limitation where VSCode controls the lifetime of the document. Since they control the life time we can simply nuke its contents to make it seem as though it isn't open.
- Added tests to verify the new content providers allow untracked document content retrievals.

#344

### CODE REVIEW

File: src/Microsoft.AspNetCore.Razor.VSCode/src/CSharp/CSharpProjectedDocumentContentProvider.ts [hunk 1]
```diff
@@ -3,28 +3,33 @@
...
  * ------------------------------------------------------------------------------------------ */
 
-import * as vscode from 'vscode';
+import { IEventEmitterFactory } from '../IEventEmitterFactory';
 import { IRazorDocumentChangeEvent } from '../IRazorDocumentChangeEvent';
+import { IRazorDocumentManager } from '../IRazorDocumentManager';
 import { RazorDocumentChangeKind } from '../RazorDocumentChangeKind';
-import { RazorDocumentManager } from '../RazorDocumentManager';
 import { getUriPath } from '../UriPaths';
+import * as vscode from '../vscodeAdapter';
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.VSCode/src/CSharp/CSharpProjectedDocumentContentProvider.ts]: The extra changes to this document are to enable it to be tested. Essentially moved it and the Html variant over to use the vscodeAdapter so we can unit test it properly.

File: src/Microsoft.AspNetCore.Razor.VSCode/src/extension.ts [hunk 2]
```diff
@@ -105,6 +105,7 @@ export async function activate(context: ExtensionContext, languageServerDir: str
 
         context.subscriptions.push(languageServerClient, onStartRegistration, onStopRegistration, logger);
     } catch (error) {
+        logger.logError('Failed when activating Razor VSCode.', error);
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.VSCode/src/extension.ts]: When moving the projected document providers to the vscode adapter I realized we weren't reporting activation errors to the Razor log. Added this bit as a bonus.

File: src/Microsoft.AspNetCore.Razor.VSCode/src/vscodeAdapter.ts [hunk 3]
```diff
@@ -4,7 +4,40 @@
  *--------------------------------------------------------------------------------------------*/
 /* tslint:disable */
 
+export type ProviderResult<T> = T | undefined | null | Thenable<T | undefined | null>;
```
Comment (NTaylorMullen) [inline src/Microsoft.AspNetCore.Razor.VSCode/src/vscodeAdapter.ts]: All of this is pulled from vscode's definition file

File: src/Microsoft.AspNetCore.Razor.VSCode/src/CSharp/CSharpProjectedDocumentContentProvider.ts [hunk 4]
```diff
@@ -3,28 +3,33 @@
...
     }
 
     public get onDidChange() { return this.onDidChangeEmitter.event; }
 
     public async provideTextDocumentContent(uri: vscode.Uri) {
         const razorDocument = this.findRazorDocument(uri);
         if (!razorDocument) {
-            vscode.window.showErrorMessage('For some reason the projected document isn\'t available.');
-            return;
+            // Document was removed from the document manager, meaning there's no more content for this
```
Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Razor.VSCode/src/CSharp/CSharpProjectedDocumentContentProvider.ts]: Is there no other way to workaround this? Doesn't seem super right to not report an error when a document is missing.

Comment (NTaylorMullen) [reply src/Microsoft.AspNetCore.Razor.VSCode/src/CSharp/CSharpProjectedDocumentContentProvider.ts]: The only way would be to maintain a list of remove documents. Hmm, I'll add a trace level log here so if we encounter errors we can at least determine what happened.

Comment (ajaybhargavb) [reply src/Microsoft.AspNetCore.Razor.VSCode/src/CSharp/CSharpProjectedDocumentContentProvider.ts]: Sounds good 👍

**Pull request merged**